community/cypher/ir-3.4/src/main/scala/org/neo4j/cypher/internal/ir/v3_4/QueryGraph.scala

```diff
@@ -218,33 +177,32 @@ case class QueryGraph(patternRelationships: Set[PatternRelationship] = Set.empty
   def allKnownRelProperties: Set[Property] =
     patternRelationships.map(_.name).flatMap(knownProperties) ++ optionalMatches.flatMap(_.allKnownRelProperties)
 
-
   def findRelationshipsEndingOn(id: IdName): Set[PatternRelationship] =
     patternRelationships.filter { r => r.left == id || r.right == id }
 
   def allPatternRelationships: Set[PatternRelationship] =
     patternRelationships ++ optionalMatches.flatMap(_.allPatternRelationships) ++
-    // Recursively add relationships from the merge-read-part
-    mergeNodePatterns.flatMap(_.matchGraph.allPatternRelationships) ++
-    mergeRelationshipPatterns.flatMap(_.matchGraph.allPatternRelationships)
+      // Recursively add relationships from the merge-read-part
+      mergeNodePatterns.flatMap(_.matchGraph.allPatternRelationships) ++
+      mergeRelationshipPatterns.flatMap(_.matchGraph.allPatternRelationships)
 
-  def coveredIdsExceptArguments: Set[IdName] = {
-    val patternIds = QueryGraph.coveredIdsForPatterns(patternNodes, patternRelationships)
-    patternIds ++ selections.predicates.flatMap(_.dependencies)
-  }
-
-  def coveredIds: Set[IdName] = {
-    val patternIds = QueryGraph.coveredIdsForPatterns(patternNodes, patternRelationships)
-    patternIds ++ argumentIds ++ selections.predicates.flatMap(_.dependencies)
-  }
+  /**
+    * Variables are bound after matching this QG, but before optional
+    * matches and updates have been applied
+    */
+  def idsWithoutOptionalMatchesOrUpdates: Set[IdName] =
+    QueryGraph.coveredIdsForPatterns(patternNodes, patternRelationships) ++ argumentIds
```

Comment:
Just to understand better, why was `selections.predicates.flatMap(_.dependencies)` removed?

Reply:
The expression that was removed was actually not listing variables exposed, but instead which dependencies the predicates have. I changed https://github.com/neo4j/neo4j/pull/10533/files#diff-0ef12799d6f97a3d6751a5ced882c8b9R81 to use the dependencies explicitly. That is also why there are so many failures - we relied on this weirdness in a bunch of places. I'll add at least one more commit to this PR